cajita/unit_test/tstParticleGridDistributor2d.hpp

```diff
@@ -245,6 +245,72 @@ void localOnlyTest( const GridType global_grid, const double cell_size )
         }
 }
 
+//---------------------------------------------------------------------------//
+// The objective of this test is to verify that particles outside the grid are
+// removed correctly if outside the global domain. In this case no particles
+// should remain in the system since they are all created outside (destination
+// rank is set to -1 if not inside the system).
+template <class GridType>
+void removeOutsideTest( const GridType global_grid )
+{
+    // Get the local block with a halo of 2.
+    const int data_halo_size = 2;
+    auto block = Cajita::createLocalGrid( global_grid, data_halo_size );
+    auto local_mesh = Cajita::createLocalMesh<Kokkos::HostSpace>( *block );
+
+    // Allocate particles
+    auto owned_cell_space =
+        block->indexSpace( Cajita::Own(), Cajita::Cell(), Cajita::Local() );
+    int num_particle = owned_cell_space.size();
+    using MemberTypes = Cabana::MemberTypes<double[2], int>;
+    using ParticleContainer = Cabana::AoSoA<MemberTypes, Kokkos::HostSpace>;
+    ParticleContainer particles( "particles", num_particle );
+    auto coords = Cabana::slice<0>( particles, "coords" );
+
+    // Put particles outside the global boundary in each dimension.
+    int pid = 0;
+    for ( int j = 0; j < owned_cell_space.extent( Dim::J ); ++j )
+        for ( int i = 0; i < owned_cell_space.extent( Dim::I ); ++i )
+        {
+            // Set the coordinates at the cell center - all outside domain.
+            bool low_i = global_grid->onLowBoundary( Dim::I );
+            bool low_j = global_grid->onLowBoundary( Dim::J );
+            bool high_i = global_grid->onHighBoundary( Dim::I );
+            bool high_j = global_grid->onHighBoundary( Dim::J );
+            if ( low_i || low_j || high_i || high_j )
+            {
+                double i_corner = local_mesh.lowCorner( Cajita::Own(), Dim::I );
+                double i_extent = local_mesh.extent( Cajita::Own(), Dim::I );
+                double j_corner = local_mesh.lowCorner( Cajita::Own(), Dim::J );
+                double j_extent = local_mesh.extent( Cajita::Own(), Dim::J );
+                if ( low_i )
+                    coords( pid, Dim::I ) = i_corner - 2 * i_extent;
```

Comment:
You could avoid having to do this by looping over the ghosted cell space. If the domain is on a low boundary, use the low corner coordinate for the particle, if it is on a high boundary, use the high corner coordinate for the particle where the corner coordinates are for the ghosted local mesh. That avoids this logic and you only make a few particles on the domain corners.

Reply:
Yeah I agree. I used the boundary indices to create the particles and I think it's more clear